We can now use the "addOffset" function with negative offsets.

lemire opened the pull request:
The implementation is simpler and relies on binary searches repeatedly (for simplicity)
so the performance should be lower than the previous implementation.

Note that we preserve the cardinality, so the offset can wrap values around.

This would fix https://github.com/RoaringBitmap/RoaringBitmap/issues/331

lemire commented:
I would prefer to get one independent reviewer to check things out before I merge.

richardstartin commented:
@lemire I'll do this tomorrow evening BST unless someone else wants to jump in.

ppiotrow commented on roaringbitmap/src/main/java/org/roaringbitmap/RoaringBitmap.java:
```diff
@@ -192,59 +192,69 @@ private static void rangeSanityCheck(final long rangeStart, final long rangeEnd)
-      RoaringBitmap answer = x.clone();
-      for(int pos = 0; pos < answer.highLowContainer.size(); pos++) {
-        int key = Util.toIntUnsigned(answer.highLowContainer.getKeyAtIndex(pos));
+      RoaringBitmap answer = new RoaringBitmap();
```
We may preallocate answer's `RoaringArray` here as we know the size in advance.

ppiotrow replied:
RoaringArray may still be resized in this case.
What do you think about 
```java
//RoaringBitmap.java
RoaringArray ra = x.highLowContainer.clone().iAddOffset(container_offset)
// or  x.highLowContainer.addOffset(container_offset)
return new RoaringBitmap(ra)
```
```java
//RoaringArray.java
public iAddOffset(int containerOffset) {
  for (int i=0;k<this.size;i++) {
    int key = ...;
    key+= containerOffset
    if((key >= 0) || (key <= 0xFFFF))  {
      keys[i]=key
    }
  }
}
```
This approach moves the responsibility of managing containers array into RoaringArray class. Clone gives us correctly preallocated arrays.

richardstartin replied:
But it's not correct. If I have the values 98303 and 98305 in a bitmap, if I shift left 2, I have the same number of keys. If I shift left by 65536, I have the same number of keys, and I can just subtract 1 from the existing key. If I shift left 32768, I now have 2 keys with 1 bit in each container. If the shift saturates rather than wraps, we could have far fewer keys than in the input.

ppiotrow replied:
Not sure if I get your point @richardstartin . This branch of code shifts **only** by multiple of 2^16. This means we always have the same number of containers (excluding scenario when containers are discarded because of moving out of range).

richardstartin replied:
Sorry, I wrote that on a phone and didn’t scroll up to the of statement. I agree. My point was that you can’t do it in the general case, only when the offset divides 2^16. If you move it to roaring array you should move the divisibility check too.

richardstartin commented:
Looks good to me but just a few thoughts

1) This method changes behaviour, without fixing a bug. The method didn't used to wrap, but now does.
2) Is wrapping what a user would want? Or would they prefer saturation?
3) As you note, this can perform better because you could apply the shift in constant time as you iterate over the keys.
4) As @ppiotrow the number of output keys can be calculated in every case, allowing pre-allocation. Based on my observations, resizing RoaringArrays for large bitmaos can cause the kind of allocation that fills up TLABs, causing GC events which result in other objects getting retained, without short[] being the obvious culprit.

lemire commented:
> This method changes behaviour, without fixing a bug.

True.

> The method didn't used to wrap, but now does.

True.

> Is wrapping what a user would want? Or would they prefer saturation?

It would be massively easier to implement without the warping around.

lemire commented:
Current status: this will not be merged as-is.

lemire commented:
@richardstartin 

Ok. So I changed the implementation, no warp around, but possible data loss if the values exceed the range.

@ppiotrow  With the new implementation, we do appends... that should be efficient.

ppiotrow commented on roaringbitmap/src/main/java/org/roaringbitmap/RoaringBitmap.java:
```diff
@@ -191,44 +191,56 @@ private static void rangeSanityCheck(final long rangeStart, final long rangeEnd)
+  public static RoaringBitmap addOffset(final RoaringBitmap x, long offset) {
+    // we need "offset" to be a long because we want to support values
+    // between -0xFFFFFFFF up to +-0xFFFFFFFF
+    long container_offset_long = offset < 0 
```
why not camelCase?

richardstartin commented:
@lemire is it worth figuring out if anyone is using this method before changing its semantics? There is an open issue for implementing shifts. Two new methods shiftLeft and shiftRight could be added, and they would "saturate" just as << and >>> do. Methods rotateLeft and rotateRight could wrap.

lemire commented:
@ppiotrow I would not mind change internals to camel case but I suspect that the internal code is not currently consistent and camel case is not yet a style requirement.

lemire commented:
@richardstartin 

All good points. Want to take a stab at a better API? What I have now should be correct.

lemire commented:
Anyhow, you guys should feel free to do better than I did... I think what I did is functional, but if you can do better, please do.

richardstartin commented:
@lemire I think the code is fine. Only concern is behaviour change in a function I'm fairly sure is in use.

lemire commented:
@richardstartin Ok. So let us discuss that. What is your concern?

If someone has been using this function with negative offsets, they are obviously out of luck, but whatever they were getting is incorrect.

If they were using this function with offset too high, they were getting exceptions. Now, they lose data points (which they can verify by a cardinality check). That's a substantive change, and I can easily revert it back to throwing exceptions.

I seriously doubt anyone was using this function and checking the exceptions. 

So I submit to you that the new behaviour won't affect anyone.

Why not keep throwing exceptions? In my opinion, that was not a good design. For one thing, it was not documented.  It was kind of hard for the end-user to know whether exceptions would get thrown. And if this ever happened, their app would crash.

The new behaviour is not perfect but seems more reasonable.

I am hoping to changing the behaviour.

richardstartin commented:
@lemire I'm talking about wrapping. There used to be a noisy guard against bugs that developers would notice, now there is silent but surprising behaviour. I haven't used the method so don't have much skin in the game - you certainly don't need my permission to merge this.

richardstartin commented:
I'm also wary of performance for performance's sake. The impetus for this PR was that someone asked about negative offsets, and it's great to add that (broken record: the issue wasn't about wanting to wrap). If performance becomes a problem, someone would submit a PR to fix it.

lemire commented:
> I'm also wary of performance for performance's sake. 

That's not why I reverted the warping around. I reverted this change because I felt you had argued that this was too much of a change.

Ok.

So what about this... what about adding back exceptions? And *then*... adding a "addOffsetSaturated". Then we would have the two behaviour.

richardstartin commented:
> That's not why I reverted the warping around. I reverted this change because I felt you had argued that this was too much of a change.

No I mean there's no reason not to merge this because it could be optimised. If this gets used a lot someone would fix the performance. I don't think we need to change the method as it is, it's more useful now.

lemire commented:
Ok. 

@ppiotrow : I am going to merge this. I think you have ideas on how to improve the performance further. I think that performance could be improved in separate pull requests.

Merging.
